components/tools/OmeroWeb/omeroweb/webgateway/static/webgateway/js/ome.roidisplay.js

```diff
@@ -403,11 +426,215 @@ $.fn.roi_display = function(options) {
+                            return false;
+                        }
+                    }
+                }
+            }
+            return true;
+        }
+
+        // Check if there is another shape with on the same Z and T planes for this ROI
+        var check_ext_shape_planes = function(roi_id, shape_z, shape_t) {
+            for (var rx=0; rx<external_roi.length; rx++) {
+                if (external_roi[rx]["id"] == roi_id) {
+                    var shapes = external_roi[rx]["shapes"];
+                    for(var sx=0; sx<shapes.length; sx++) {
+                        if (shapes[sx]["theZ"] == shape_z && shapes[sx]["theT"] == shape_t) {
+                            console.error("Z plane " + shape_z + " and T plane " + shape_t + " already used");
+                            return false;
+                        }
+                    }
+                }
+            }
+            return true;
+        }
+
+        // configure a shape obtained using function in ome.roiutils.js as a proper shape for OMERO.web viewer
+        var configure_shape = function(shape_id, shape_config) {
+            shape_config["id"] = shape_id;
+        }
+
+        // build a proper ROI definition that can be used by OMERO.web viewer
+        var build_roi_description = function(roi_id, shapes) {
+            return {
+                "id": roi_id,
+                "shapes": typeof shapes !== "undefined" ? shapes : []
+            };
+        }
+
+        // append a shape to ROI with ID roi_id
+        var append_shape = function(roi_id, shape_config, rois_collection) {
+            for (var x=0; x<rois_collection.length; x++) {
+                if (rois_collection[x]["id"] == roi_id) {
+                    // ROI with ID roi_id already exists, append shape
+                    rois_collection[x]["shapes"].push(shape_config);
+                    return;
+                }
+            }
+            var roi = build_roi_description(roi_id, [shape_config]);
+            rois_collection.push(roi);
+        };
+
+        /*
+        Add a SHAPE definition to OMERO.web viewer. Shape will be associated to ROI with ID roi_id
+        and will obtain ID shape_id. Shape is described using shape_config dictionary.
+        If refresh_rois is TRUE, automatically refresh viewport.
+        If hide_ome_rois is TRUE, hide OME ROI when refreshing viewport.
+         */
+        this.push_shape = function(roi_id, shape_id, shape_config, refresh_rois, hide_ome_rois) {
+            if (roi_json == null) {
+                load_rois(false, undefined, roi_id, shape_id, shape_config, refresh_rois, hide_ome_rois);
+            } else {
+                $viewportimg.trigger("push_shape", [roi_id, shape_id, shape_config, refresh_rois,
+                                                    hide_ome_rois]);
+            }
+        };
+
+        $viewportimg.on("push_shape", function (event, roi_id, shape_id, shape_config, refresh_rois,
```

Comment:
Is there a reason/use-case for exposing this functionality on the $viewportimg element via a trigger instead of simply moving this to a 'private' function like ``` this._push_shape ```?